packages/driver-azure/src/driver/client.ts

```diff
@@ -0,0 +1,238 @@
+  .substring(0, 16)}${randomBytes(16)
+  .toString('hex')
+  .substring(0, 8)}`
+
+export const getTags = (profileId: string, envId: string) => ({
+  [AzureCustomTags.PROFILE_ID]: profileId,
+  [AzureCustomTags.ENV_ID]: envId,
+})
+export const client = ({
+  region,
+  subscriptionId,
+  profileId,
+}: {
+  region: string
+  subscriptionId: string
+  profileId: string
+}) => {
+  const credentials = new DefaultAzureCredential()
+  const resourceClient = new ResourceManagementClient(
+    credentials,
+    subscriptionId
+  )
+  const computeClient = new ComputeManagementClient(credentials, subscriptionId)
+  const storageClient = new StorageManagementClient(credentials, subscriptionId)
+  const networkClient = new NetworkManagementClient(credentials, subscriptionId)
+  return {
+    deleteResourcesResourceGroup: async (resourceGroupName: string, wait: boolean) => {
+      if (wait) {
+        await resourceClient.resourceGroups.beginDeleteAndWait(resourceGroupName, {
+          forceDeletionTypes: 'Microsoft.Compute/virtualMachines,Microsoft.Compute/virtualMachineScaleSets',
+        })
+      }
+      await resourceClient.resourceGroups.beginDelete(resourceGroupName, {
+        forceDeletionTypes: 'Microsoft.Compute/virtualMachines,Microsoft.Compute/virtualMachineScaleSets',
+      })
+    },
+    listResourceGroups: (): AsyncIterableIterator<ResourceGroup> => {
+      const filter = `tagName eq '${AzureCustomTags.PROFILE_ID}' and tagValue eq '${profileId}'`
+      return resourceClient.resourceGroups.list({ filter })
+    },
+    listResource: (): AsyncIterableIterator<GenericResource> => {
+      const filter = `tagName eq '${AzureCustomTags.PROFILE_ID}' and tagValue eq '${profileId}'`
+      return resourceClient.resources.list({ filter })
+    },
+    listInstances: (): AsyncIterableIterator<VMInstance> => asyncMap(
+      async vm => (
+        { vm, ...(await getIpAddresses(networkClient, vm)) }),
+      asyncFilter<VirtualMachine>(({ tags, provisioningState }) =>
+        tags?.[AzureCustomTags.PROFILE_ID] === profileId && provisioningState !== 'Deleting', computeClient.virtualMachines.listAll())
+    ),
+    getInstance: async (envId: string): Promise<VMInstance> => {
+      const filter = `tagName eq '${AzureCustomTags.ENV_ID}' and tagValue eq '${envId}'`
+      const vmResource = await asyncFirst(asyncFilter(x => x.type === 'Microsoft.Compute/virtualMachines', resourceClient.resources.list({ filter })))
+      if (!vmResource?.id) {
+        throw new AzureResourceNotFound(`envId: ${envId}`)
+      }
+      const name = extractResourceNameFromId(vmResource.id)
+      const resourceGroupName = extractResourceGroupNameFromId(vmResource.id)
+      const vm = await computeClient.virtualMachines.get(resourceGroupName, name)
+      if (vm.provisioningState === 'Deleting') {
+        throw new AzureResourceNotFound(`current vm instance for envId: ${envId} is deleting`)
```

Comment:
error, the wrapping func is looking if it's 404 and returns unknown if needed.
this is an edge case since we still have a VM but it is being deleted.